HDFS-15305. Extend ViewFS and provide ViewFSOverloadScheme implementation with scheme configurable.

umamaheswararao opened the pull request:
https://issues.apache.org/jira/browse/HDFS-15305

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 34s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 5 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m  4s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  20m 12s |  trunk passed  |
| +1 :green_heart: |  compile  |  17m 36s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   2m 40s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 56s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  20m 18s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  4s |  trunk passed  |
| +0 :ok: |  spotbugs  |   3m  5s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 10s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m  0s |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 18s |  the patch passed  |
| +1 :green_heart: |  javac  |  16m 18s |  the patch passed  |
| -0 :warning: |  checkstyle  |   2m 41s |  root: The patch generated 13 new + 130 unchanged - 1 fixed = 143 total (was 131)  |
| +1 :green_heart: |  mvnsite  |   2m 55s |  the patch passed  |
| -1 :x: |  whitespace  |   0m  0s |  The patch has 10 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| +1 :green_heart: |  shadedclient  |  13m 58s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  1s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   5m 22s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   9m 12s |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  |  93m 11s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  6s |  The patch does not generate ASF License warnings.  |
|  |   | 220m 21s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.io.compress.snappy.TestSnappyCompressorDecompressor |
|   | hadoop.io.compress.TestCompressorDecompressor |
|   | hadoop.hdfs.server.datanode.TestBPOfferService |
|   | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |
|   | hadoop.TestRefreshCallQueue |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1988 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 35e0641a50f8 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / b5b45c53a4e |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/1/artifact/out/diff-checkstyle-root.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/1/artifact/out/whitespace-eol.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/1/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/1/testReport/ |
| Max. process+thread count | 5007 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/1/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

umamaheswararao commented:
Just a note: I have not verified nfly mount links in this patch. I will file another JIRA for nfly changes and tests.
Nfly using FileSystem.get, so I should make nfly also use FsCreator to handle looping.

virajith commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  @Override
+  public void initialize(final URI theUri, final Configuration conf)
```
If we add a protected method `getFsCreator(String scheme)` in `ViewFileSystem` and override it this class,I think `initialize` can be made much simpler. it just needs to `conf.setBoolean(CONFIG_VIEWFS_ENABLE_INNER_CACHE, true)` and call  `super.initialize(theUri, conf)`. Is that correct?

umamaheswararao replied:
Thanks. Refactored and please check if it is addressing this comment.

umamaheswararao commented:
Thank you @virajith for reviewing the PR. I have pushed another commit by addressing all of your comments. Refactored a bit about initialization part which fixes most of the comments.
Fixed the check style and whitespace issues as well.
Please take a look.

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FsConstants.java:
```diff
@@ -42,4 +42,11 @@
+  public static final String VIEWFS_OVERLOAD_SCHEME_DEFAULT = "hdfs";
+  public static final String FS_VIEWFS_OVERLOAD_SCHEME_TARGET_FS_IMPL_PATTERN_KEY =
+      "fs.viewfs.overload.scheme.target.%s.impl";
+  public static final String FS_IMPL_PATTERN_KEY = "fs.%s.impl";
```
Initial commit was using first two in src, anyway now I removed config of VIEWFS_OVERLOAD_SCHEME_DEFAULT completely. So, only FS_VIEWFS_OVERLOAD_SCHEME_TARGET_FS_IMPL_PATTERN is there now and using this in ViewFsOverloadScheme class.

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java:
```diff
@@ -302,6 +320,11 @@ protected FileSystem getTargetFileSystem(final String settings,
     }
   }
 
+  protected void superFSInit(final URI theUri, final Configuration conf)
```
Removed this method now. After refactoring initialization part, this got addressed along with that.

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,175 @@
+
+/******************************************************************************
+ * This class is extended from the ViewFileSystem for the overloaded scheme 
+ * file system. This object is the way end-user code interacts with a multiple
```
Thanks corrected them.

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,175 @@
+import org.apache.hadoop.fs.FsConstants;
+import org.apache.hadoop.fs.UnsupportedFileSystemException;
+
+/******************************************************************************
```
I added examples please let me know if they are make sense.

umamaheswararao replied:
#b) added config details. Further more details, we will be covering in user guide.
#a) first point, yes that ViewFS function. second part, we have overridden getScheme, and that will allow VFSOS to work with any scheme. Any scheme flexibility should come to this class only.

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,175 @@
+    superFSInit(theUri, conf);
+    setConf(conf);
+    config = conf;
+    myScheme = config.get(FsConstants.VIEWFS_OVERLOAD_SCHEME_KEY);
```
Good point. This config is not required now. Allowing to init by any scheme URI.

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,175 @@
+    superFSInit(theUri, conf);
+    setConf(conf);
+    config = conf;
+    myScheme = config.get(FsConstants.VIEWFS_OVERLOAD_SCHEME_KEY);
```
same as above. Removed this now and that should address your comment. Please let me know if that not the case.

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,175 @@
+
+      /**
+       * This method is overridden because in ViewFsOverloadScheme if
+       * overloaded scheme matches with mounted target fs scheme, file system
```
corrected. Thanks

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,175 @@
+      @Override
+      public FileSystem createFs(URI uri, Configuration conf)
+          throws IOException {
+        if (uri.getScheme().equals(myScheme)) {
```
Yes, addressed this. Thanks

virajith replied:
Don't think this is addressed. I am saying `createFileSystem` should be called here for every scheme for which the appropriate FS_VIEWFS_OVERLOAD_SCHEME_TARGET_FS_IMPL_PATTERN_KEY is set and only those where `uri.getScheme().equals(myScheme)` . Wdyt?

umamaheswararao replied:
For other schemes which are not matching will be able to get it resolved by FileSystem itself right?
When target uri is scheme is same as OverloadScheme uri sheme, then only we need to bypass to avoid looping. Other cases, FileSystem#get or FileSystem#newInstance will get the right instance. Am I missing?
Since your exposed uri scheme occupied with OverloadScheme impl, we need that additional config to get actual impl. This was the idea of that configuration.

>Currently, the override only works for one scheme.  
User is going to use one scheme from out side right per OverloadScheme instance right?
Because your OverloadScheme init will take one uri to initialize, That will be user initialized uri. Rest all will go as mapping uris. 

example: inited uri is hdfs://xyz:9000
                                    mapping target uris are /target1 -> hdfs://target1:9000, /target2 -> hdfs://target2:9000, /target3 -> s3a://bucket1/target3

Here when getting target fs: we would check if scheme matches with inited. Here hdfs was inited. So, for hdfs scheme, FileSystem#get resolution will always to OverloadSchem impl. To get actual impl we use the FS_VIEWFS_OVERLOAD_SCHEME_TARGET_FS_IMPL_PATTERN_KEY to initialize target fs. Other impls should be able to resolved by FileSystem#get.
Is your use case is for configuring Overload scheme for all fs.<scheme>.impl in same client process? example in same client process, you would configure, fs.hdfs.impl=ViewFSOverloadScheme and fs.s3a.impl=ViewFSOverloadScheme ?

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,175 @@
+       * This method is overridden because in ViewFsOverloadScheme if
+       * overloaded scheme matches with mounted target fs scheme, file system
+       * should be created without going into fs.<scheme>.impl based 
+       * resolution. Otherwise it will end up into loop as target will be 
```
done.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m 22s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 5 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m  7s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  26m  2s |  trunk passed  |
| +1 :green_heart: |  compile  |  22m  2s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   3m 24s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   3m 14s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  23m 50s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  6s |  trunk passed  |
| +0 :ok: |  spotbugs  |   3m 34s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   6m 18s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 18s |  the patch passed  |
| -1 :x: |  compile  |  22m 30s |  root in the patch failed.  |
| -1 :x: |  javac  |  22m 30s |  root in the patch failed.  |
| -0 :warning: |  checkstyle  |   3m 55s |  root: The patch generated 1 new + 130 unchanged - 1 fixed = 131 total (was 131)  |
| +1 :green_heart: |  mvnsite  |   3m 27s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  17m  7s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m 43s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   7m 10s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  10m 14s |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  | 118m 45s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 58s |  The patch does not generate ASF License warnings.  |
|  |   | 276m 28s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.io.compress.TestCompressorDecompressor |
|   | hadoop.io.compress.snappy.TestSnappyCompressorDecompressor |
|   | hadoop.TestRefreshCallQueue |
|   | hadoop.hdfs.server.balancer.TestBalancer |
|   | hadoop.hdfs.TestLeaseRecovery2 |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1988 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 05ee4f0f234f 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 82343790eeb |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/2/artifact/out/patch-compile-root.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/2/artifact/out/patch-compile-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/2/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/2/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/2/testReport/ |
| Max. process+thread count | 2870 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/2/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

chliang71 commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsOverloadSchemeLocalFileSystem.java:
```diff
@@ -0,0 +1,158 @@
+   */
+  @Test
+  public void testLocalFsLinkSlashMerge() throws Exception {
+    LOG.info("Starting testLocalFSCreateAndDelete");
```
shouldn't this message be testLocalFsLinkSlashMerge? similar for the other test below

umamaheswararao replied:
thanks. I will change them in next patch.

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java:
```diff
@@ -96,16 +96,49 @@ static AccessControlException readOnlyMountTable(final String operation,
+  /**
+   * Gets file system creator instance.
+   */
+  protected FsGetter fsGetter() {
```
Thanks @chliang71  for the review. We have overridden this method in ViewFSOverloadScheme class. In ViewFSOverloadScheme, fsGetter gets its special FsGetter anonymous subclass instance with its own implemented methods. Does this answers your question?

chliang71 replied:
I see, thanks!

umamaheswararao commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,170 @@
+ *****************************************************************************/
+@InterfaceAudience.LimitedPrivate({ "MapReduce", "HBase", "Hive" })
+@InterfaceStability.Evolving
+public class ViewFsOverloadScheme extends ViewFileSystem {
```
Sure I will add logging for this.

umamaheswararao commented:
@chliang71 Thank you for the reviews. I have pushed a PR by addressing your comments.
Please take a look if it addresses your concerns.

-Thank you!

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  22m 15s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 5 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 48s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  18m 49s |  trunk passed  |
| +1 :green_heart: |  compile  |  17m  0s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   2m 42s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 57s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  20m  5s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  4s |  trunk passed  |
| +0 :ok: |  spotbugs  |   3m  5s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 11s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 57s |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 25s |  the patch passed  |
| +1 :green_heart: |  javac  |  16m 25s |  the patch passed  |
| -0 :warning: |  checkstyle  |   3m  4s |  root: The patch generated 1 new + 130 unchanged - 1 fixed = 131 total (was 131)  |
| +1 :green_heart: |  mvnsite  |   2m 58s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  13m 55s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  4s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   5m 25s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   9m 13s |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  |  94m 32s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  5s |  The patch does not generate ASF License warnings.  |
|  |   | 241m 21s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.io.compress.snappy.TestSnappyCompressorDecompressor |
|   | hadoop.io.compress.TestCompressorDecompressor |
|   | hadoop.hdfs.TestDecommissionWithBackoffMonitor |
|   | hadoop.hdfs.TestDeadNodeDetection |
|   | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |
|   | hadoop.TestRefreshCallQueue |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1988 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 8e953de40c64 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 82343790eeb |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/3/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/3/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/3/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/3/testReport/ |
| Max. process+thread count | 4209 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/3/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

virajith commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsOverloadSchemeHdfsFileSystemContract.java:
```diff
@@ -0,0 +1,122 @@
+    ConfigUtil.addLink(conf, defaultFSURI.getAuthority(),
+        "/FileSystemContractBaseTest/",
+        new URI(defaultFSURI.toString() + "/FileSystemContractBaseTest/"));
+    fs = (ViewFsOverloadScheme) FileSystem.get(conf);
```
cast not required.

virajith commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsOverloadSchemeLocalFileSystem.java:
```diff
@@ -0,0 +1,158 @@
+    fileSystemTestHelper = new FileSystemTestHelper();
+    // create the test root on local_fs
+    targetTestRoot = fileSystemTestHelper.getAbsoluteTestRootPath(fsTarget);
+    fsTarget.delete(targetTestRoot, true);
```
Except for these two, can the rest be done `@BeforeClass`?

umamaheswararao replied:
Some tests has slashmerge config which will not work if there are other mount links in config. I think its ok to keep config inits in @Before ( local fs is not heavy)

virajith commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsOverloadSchemeHdfsFileSystemContract.java:
```diff
@@ -0,0 +1,122 @@
+    conf.set(CommonConfigurationKeys.FS_PERMISSIONS_UMASK_KEY,
+        FileSystemContractBaseTest.TEST_UMASK);
+    final File basedir = GenericTestUtils.getRandomizedTestDir();
+    cluster = new MiniDFSCluster.Builder(conf, basedir)
```
do we need the `cluster` created for every test? seems heavy :| 

Also, can we set `FS_DEFAULT_NAME_KEY` explicitly here?

umamaheswararao replied:
Let me check that if I can start cluster once and use. We reused existing HDFSContract test.
If existing contract tests passes with one time cluster start, I should be good to do that. I will try that.
Why do we need to set  FS_DEFAULT_NAME_KEY explicitly? Minicluster will do that for is we use that same config.

umamaheswararao replied:
I tried to moved cluster init into Before class. Contract tests are passing. Please check.

virajith commented on hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFsOverloadScheme.java:
```diff
@@ -0,0 +1,187 @@
+ *****************************************************************************/
+@InterfaceAudience.LimitedPrivate({ "MapReduce", "HBase", "Hive" })
+@InterfaceStability.Evolving
+public class ViewFsOverloadScheme extends ViewFileSystem {
```
I think `ViewOverloadSchemeFilesystem` is a better name. Also do we need an implementation extending `ViewFs` class (which extends `AbstractFileSystem`)?

umamaheswararao replied:
This we discussed. Sanjay suggested to keep ViewFsOverloadScheme to indicate that most of the implementation is ViewFS itself but we get to overload the scheme. I can this if there are no objections anyway. ViewOverloadSchemeFilesystem and ViewFsOverloadScheme are discussed names while writing the doc. :-)

Yes, we have that FileContext based implementation in consideration. I will be filing a JIRA for that.

virajith commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsOverloadSchemeWithHdfsScheme.java:
```diff
@@ -0,0 +1,351 @@
+    final Path hdfsTargetPath = new Path(defaultFSURI + HDFS_USER_FOLDER);
+    final Path localTragetPath = new Path(localTargetDir.toURI());
+    conf = new Configuration();
+    createLinks(true, hdfsTargetPath, localTragetPath);
```
calling createLinks isn't even needed here right?

umamaheswararao replied:
Without creating links ViewFS fails with "not mount points configured". So, to move forward to the intended scenario, we just need to create links.

virajith commented on hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsOverloadSchemeWithHdfsScheme.java:
```diff
@@ -0,0 +1,351 @@
+   * It should be able to create file using ViewFsOverloadScheme.
+   */
+  @Test(timeout = 30000)
+  public void testViewFsOverloadSchemeWhenInnerCacheDisabled()
```
Can we add a test that cache actually works?

umamaheswararao replied:
We have not changed cache behavior itself. When cache disabled, we have varied path. 
Anyway Its should be easy to add a test for that I think. Let me add one.

umamaheswararao replied:
Added Tests.

umamaheswararao commented on hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsOverloadSchemeLocalFileSystem.java:
```diff
@@ -0,0 +1,158 @@
+    final FileSystem lViewFS = FileSystem.get(mountURI, conf);
+    try {
+      Path testPath = new Path(mountURI.toString() + "/lfsroot/test");
+      lViewFS.create(testPath);
```
Let me use this other method which closes itself. I dont need stream here.

umamaheswararao commented:
@virajith Thank you for your time on reviews. This commit should address most of your comments. Additionally I fixed an issue when getTragetFileSystem should use passed uri instead of initialized uri. I have corrected that.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  26m 34s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 5 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 52s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m 54s |  trunk passed  |
| +1 :green_heart: |  compile  |  18m 10s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   2m 51s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 48s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  21m 22s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m 46s |  trunk passed  |
| +0 :ok: |  spotbugs  |   3m 12s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 19s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 26s |  the patch passed  |
| +1 :green_heart: |  compile  |  22m  5s |  the patch passed  |
| +1 :green_heart: |  javac  |  22m  5s |  the patch passed  |
| -0 :warning: |  checkstyle  |   3m  9s |  root: The patch generated 17 new + 130 unchanged - 1 fixed = 147 total (was 131)  |
| +1 :green_heart: |  mvnsite  |   2m 50s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 44s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   1m 44s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   5m 34s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   9m 30s |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  | 124m 32s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 58s |  The patch does not generate ASF License warnings.  |
|  |   | 288m 36s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.fs.viewfs.TestViewFsTrash |
|   | hadoop.io.compress.TestCompressorDecompressor |
|   | hadoop.io.compress.snappy.TestSnappyCompressorDecompressor |
|   | hadoop.hdfs.server.namenode.ha.TestBootstrapStandby |
|   | hadoop.hdfs.server.blockmanagement.TestUnderReplicatedBlocks |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1988 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 697ebdd366f4 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 44de193bec3 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/4/artifact/out/diff-checkstyle-root.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/4/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/4/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/4/testReport/ |
| Max. process+thread count | 3123 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/4/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 36s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 5 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 35s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  24m 59s |  trunk passed  |
| +1 :green_heart: |  compile  |  21m 33s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   3m 16s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   3m 11s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  23m 44s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  0s |  trunk passed  |
| +0 :ok: |  spotbugs  |   3m 43s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   6m 14s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 20s |  the patch passed  |
| +1 :green_heart: |  compile  |  20m 44s |  the patch passed  |
| +1 :green_heart: |  javac  |  20m 43s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   3m  5s |  root: The patch generated 0 new + 130 unchanged - 1 fixed = 130 total (was 131)  |
| +1 :green_heart: |  mvnsite  |   3m  7s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  17m  2s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   2m  1s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   6m 50s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  10m 29s |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  | 132m  3s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  2s |  The patch does not generate ASF License warnings.  |
|  |   | 283m 51s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.io.compress.TestCompressorDecompressor |
|   | hadoop.io.compress.snappy.TestSnappyCompressorDecompressor |
|   | hadoop.hdfs.server.namenode.ha.TestAddBlockTailing |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1988 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux a80acf9e19bc 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 8dace8ff3a9 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/5/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/5/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/5/testReport/ |
| Max. process+thread count | 3007 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/5/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m  0s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 5 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 16s |  Maven dependency ordering for branch  |
| -1 :x: |  mvninstall  |   0m 25s |  root in trunk failed.  |
| -1 :x: |  compile  |   0m 17s |  root in trunk failed.  |
| -0 :warning: |  checkstyle  |   0m 20s |  The patch fails to run checkstyle in root  |
| -1 :x: |  mvnsite  |   0m 28s |  hadoop-common in trunk failed.  |
| -1 :x: |  mvnsite  |   0m 23s |  hadoop-hdfs in trunk failed.  |
| +1 :green_heart: |  shadedclient  |   1m 30s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 16s |  hadoop-common in trunk failed.  |
| +0 :ok: |  spotbugs  |   4m 16s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| -1 :x: |  findbugs  |   0m 43s |  hadoop-common in trunk failed.  |
| -0 :warning: |  patch  |   4m 27s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 41s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m 34s |  the patch passed  |
| -1 :x: |  compile  |   2m 40s |  root in the patch failed.  |
| -1 :x: |  javac  |   2m 40s |  root in the patch failed.  |
| -0 :warning: |  checkstyle  |   3m 25s |  root: The patch generated 130 new + 0 unchanged - 0 fixed = 130 total (was 0)  |
| -1 :x: |  mvnsite  |   0m 29s |  hadoop-common in the patch failed.  |
| -1 :x: |  mvnsite  |   0m 28s |  hadoop-hdfs in the patch failed.  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| -1 :x: |  shadedclient  |   0m 31s |  patch has errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 15s |  hadoop-common in the patch failed.  |
| -1 :x: |  javadoc  |   0m 23s |  hadoop-hdfs in the patch failed.  |
| -1 :x: |  findbugs  |   0m 29s |  hadoop-common in the patch failed.  |
| -1 :x: |  findbugs  |   0m 52s |  hadoop-hdfs in the patch failed.  |
||| _ Other Tests _ |
| -1 :x: |  unit  |   0m 16s |  hadoop-common in the patch failed.  |
| -1 :x: |  unit  |   0m 25s |  hadoop-hdfs in the patch failed.  |
| -1 :x: |  asflicense  |   0m 31s |  The patch generated 1 ASF License warnings.  |
|  |   |  26m 26s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1988 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 36717d18b60a 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ebb878bab99 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/branch-mvninstall-root.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/branch-compile-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/buildtool-branch-checkstyle-root.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/branch-mvnsite-hadoop-common-project_hadoop-common.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/branch-mvnsite-hadoop-hdfs-project_hadoop-hdfs.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/branch-javadoc-hadoop-common-project_hadoop-common.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/branch-findbugs-hadoop-common-project_hadoop-common.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-compile-root.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-compile-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/diff-checkstyle-root.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-mvnsite-hadoop-common-project_hadoop-common.txt |
| mvnsite | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-mvnsite-hadoop-hdfs-project_hadoop-hdfs.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-javadoc-hadoop-common-project_hadoop-common.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-findbugs-hadoop-common-project_hadoop-common.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-findbugs-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 93 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1988/6/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

umamaheswararao commented:
Above build report seems to be due to build machine resources issues:
mvn install
......
[INFO] ------------------------------------------------------------------------
[ERROR] unable to create new native thread -> [Help 1]
[ERROR]

umamaheswararao commented:
I looked at it. In my local box mac ".Trash" is reserved. So, I changed the ./Trash to Trash1 and ran several times. It passed. It should be some flakey test. (It has some history in failing ex: HADOOP-8542)
Since it was passing in several times runs, I did not dig further.

Since recent build was having issues:
I ran locally build.
[INFO] Apache Hadoop Cloud Storage ........................ SUCCESS [  0.865 s]
[INFO] Apache Hadoop Cloud Storage Project ................ SUCCESS [  0.051 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  22:14 min
[INFO] Finished at: 2020-05-04T16:50:01-07:00
[INFO] ------------------------------------------------------------------------


[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.hadoop.fs.viewfs.TestViewFileSystemOverloadSchemeLocalFileSystem
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.48 s - in org.apache.hadoop.fs.viewfs.TestViewFileSystemOverloadSchemeLocalFileSystem
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0

[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.hadoop.fs.viewfs.TestViewFileSystemOverloadSchemeHdfsFileSystemContract
[INFO] Tests run: 44, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.732 s - in org.apache.hadoop.fs.viewfs.TestViewFileSystemOverloadSchemeHdfsFileSystemContract
[INFO] Running org.apache.hadoop.fs.viewfs.TestViewFileSystemOverloadSchemeWithHdfsScheme
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 22.184 s - in org.apache.hadoop.fs.viewfs.TestViewFileSystemOverloadSchemeWithHdfsScheme
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 56, Failures: 0, Errors: 0, Skipped: 0

[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.hadoop.fs.viewfs.TestViewFsTrash
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.412 s - in org.apache.hadoop.fs.viewfs.TestViewFsTrash
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0

umamaheswararao commented:
Thanks a lot, @virajith and @chliang71 for reviews.
